# Issue 26403: Meta-ticket: make graphs compatible with Python 3

archive/issues_026403.json:
```json
{
    "body": "CC:  @tscrim @fchapoton @jhpalmieri @jfraymond\n\nKeywords: py3, graph\n\nThis ticket is used to keep track of the progress towards python3 in graphs.\n\n\n**Major issues**\n- methods `.vertices()` and `.edges()` use sort by default\n- direct comparison of vertex labels (e.g., in method `iterator_edges` of `base/sparse_graph.pyx`)\n\n\n**Done**\n- #26621 avoid using `.vertices()` and `.edges()` in `bliss.pyx`\n\nIssue created by migration from https://trac.sagemath.org/ticket/26640\n\n",
    "created_at": "2018-11-05T10:01:16Z",
    "labels": [
        "component: group theory"
    ],
    "milestone": "https://github.com/sagemath/sagetest/milestones/sage-duplicate/invalid/wontfix",
    "title": "Meta-ticket: make graphs compatible with Python 3",
    "type": "issue",
    "url": "https://github.com/sagemath/sagetest/issues/26403",
    "user": "https://github.com/dcoudert"
}
```
CC:  @tscrim @fchapoton @jhpalmieri @jfraymond

Keywords: py3, graph

This ticket is used to keep track of the progress towards python3 in graphs.


**Major issues**
- methods `.vertices()` and `.edges()` use sort by default
- direct comparison of vertex labels (e.g., in method `iterator_edges` of `base/sparse_graph.pyx`)


**Done**
- #26621 avoid using `.vertices()` and `.edges()` in `bliss.pyx`

Issue created by migration from https://trac.sagemath.org/ticket/26640





---

archive/issue_comments_371361.json:
```json
{
    "body": "Changing component from group theory to graph theory.",
    "created_at": "2018-11-05T12:48:13Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371361",
    "user": "https://github.com/fchapoton"
}
```

Changing component from group theory to graph theory.



---

archive/issue_comments_371362.json:
```json
{
    "body": "I'm answering here the question of [#26567#comment:9](https://trac.sagemath.org/ticket/26567#comment:9) about how to make `iterator_edges` of `sparse_graph.pyx` py3 compatible, and if something similar to #26567 for `dense_graph.pyx` can be done for `sparse_graph.pyx`.\n\nSo far, I don't know what's the best option.\n- just stop ordering end vertices of edges as well as vertices in general, but it shall break many algorithms\n- change the way we use graphs to ensure that internally vertices are all integers, and then use method `get_vertex_label` (that we already have but rarely use) only when the user wants to display lists of vertices/edges. I think networkx is now doing something like that.\n- try to mimic the py2 sorting using try... except statements, but this might induce some slowdown.\n- maintain an ordering of the vertices and a mapping  from vertices to integers used when sorting list of vertices and the end vertices of edges. This ordering could be updated at vertex insertion/deletion time.\n\nAll these options have pros and cons, and each of them will require a significant amount of work to fix doctests and algorithms. \n\nIn the last months, we did significant progresses on reducing the dependency on ordering, but this is not enough and this central issue is very complex to fix. Which is the best option in the short/long term ?",
    "created_at": "2018-11-29T08:31:56Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371362",
    "user": "https://github.com/dcoudert"
}
```

I'm answering here the question of [#26567#comment:9](https://trac.sagemath.org/ticket/26567#comment:9) about how to make `iterator_edges` of `sparse_graph.pyx` py3 compatible, and if something similar to #26567 for `dense_graph.pyx` can be done for `sparse_graph.pyx`.

So far, I don't know what's the best option.
- just stop ordering end vertices of edges as well as vertices in general, but it shall break many algorithms
- change the way we use graphs to ensure that internally vertices are all integers, and then use method `get_vertex_label` (that we already have but rarely use) only when the user wants to display lists of vertices/edges. I think networkx is now doing something like that.
- try to mimic the py2 sorting using try... except statements, but this might induce some slowdown.
- maintain an ordering of the vertices and a mapping  from vertices to integers used when sorting list of vertices and the end vertices of edges. This ordering could be updated at vertex insertion/deletion time.

All these options have pros and cons, and each of them will require a significant amount of work to fix doctests and algorithms. 

In the last months, we did significant progresses on reducing the dependency on ordering, but this is not enough and this central issue is very complex to fix. Which is the best option in the short/long term ?



---

archive/issue_comments_371363.json:
```json
{
    "body": "May I ask in which ticket, if any, you do touch the \"is_isomorphic\" method ?\n\nI would really like this to work with python 3:\n\n```python\nsage: G = Graph()\nsage: G.add_edge((1,'a'))\nsage: G\nGraph on 2 vertices\nsage: G.is_isomorphic(G)\n```",
    "created_at": "2018-12-06T13:26:24Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371363",
    "user": "https://github.com/fchapoton"
}
```

May I ask in which ticket, if any, you do touch the "is_isomorphic" method ?

I would really like this to work with python 3:

```python
sage: G = Graph()
sage: G.add_edge((1,'a'))
sage: G
Graph on 2 vertices
sage: G.is_isomorphic(G)
```



---

archive/issue_comments_371364.json:
```json
{
    "body": "I have not touched any method related to isomorphism. I have only opened a ticket to show a bug with `canonical_label` #26800.",
    "created_at": "2018-12-06T13:51:49Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371364",
    "user": "https://github.com/dcoudert"
}
```

I have not touched any method related to isomorphism. I have only opened a ticket to show a bug with `canonical_label` #26800.



---

archive/issue_comments_371365.json:
```json
{
    "body": "ok.. Then either we wait for all these tickets to be closed (and this may take a lot of time) or we can try to fix this isomorphism problem now..",
    "created_at": "2018-12-06T13:53:22Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371365",
    "user": "https://github.com/fchapoton"
}
```

ok.. Then either we wait for all these tickets to be closed (and this may take a lot of time) or we can try to fix this isomorphism problem now..



---

archive/issue_comments_371366.json:
```json
{
    "body": "We can start working on it. In the worst case, I will have to fix some merge conflicts.",
    "created_at": "2018-12-06T14:03:18Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371366",
    "user": "https://github.com/dcoudert"
}
```

We can start working on it. In the worst case, I will have to fix some merge conflicts.



---

archive/issue_comments_371367.json:
```json
{
    "body": "I have made a first tentative at `u/chapoton/graphe_iso`",
    "created_at": "2018-12-06T16:59:37Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371367",
    "user": "https://github.com/fchapoton"
}
```

I have made a first tentative at `u/chapoton/graphe_iso`



---

archive/issue_comments_371368.json:
```json
{
    "body": "I'm not good enough with git to know what you have changed, except the first 2 calls to `.vertices()`. It seems ok. Note that the function has 2 more calls to `.vertices()` that are clearly useless. I think the proposed change fix some doctests, so you should open a ticket.",
    "created_at": "2018-12-06T17:28:55Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371368",
    "user": "https://github.com/dcoudert"
}
```

I'm not good enough with git to know what you have changed, except the first 2 calls to `.vertices()`. It seems ok. Note that the function has 2 more calls to `.vertices()` that are clearly useless. I think the proposed change fix some doctests, so you should open a ticket.



---

archive/issue_comments_371369.json:
```json
{
    "body": "I made #26846 for graph isomorphism",
    "created_at": "2018-12-06T19:36:40Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371369",
    "user": "https://github.com/fchapoton"
}
```

I made #26846 for graph isomorphism



---

archive/issue_comments_371370.json:
```json
{
    "body": "Salut !\n\n* we should probably add #26971 somewhere\n\n* where are the plot methods handled ?",
    "created_at": "2018-12-30T10:35:03Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371370",
    "user": "https://github.com/fchapoton"
}
```

Salut !

* we should probably add #26971 somewhere

* where are the plot methods handled ?



---

archive/issue_comments_371371.json:
```json
{
    "body": "See #26469, #26478, #26480, #26484.\n\nI will check #26971 soon.",
    "created_at": "2018-12-30T10:40:49Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371371",
    "user": "https://github.com/dcoudert"
}
```

See #26469, #26478, #26480, #26484.

I will check #26971 soon.



---

archive/issue_comments_371372.json:
```json
{
    "body": "Concerning plot, there are still problems in py3 when plotting posets. This point to the remaining\n\n```\nFile \"/home/u1/chapoton/sage3/local/lib/python3.6/site-packages/sage/graphs/graph_plot.py\", line 255, in __init__\n        self._nodelist = graph.vertices()\n```",
    "created_at": "2018-12-30T10:44:33Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371372",
    "user": "https://github.com/fchapoton"
}
```

Concerning plot, there are still problems in py3 when plotting posets. This point to the remaining

```
File "/home/u1/chapoton/sage3/local/lib/python3.6/site-packages/sage/graphs/graph_plot.py", line 255, in __init__
        self._nodelist = graph.vertices()
```



---

archive/issue_events_066077.json:
```json
{
    "actor": "https://github.com/dcoudert",
    "created_at": "2018-12-30T11:27:30Z",
    "event": "milestoned",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "milestone": "sage-8.6",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sagetest/issues/26403#event-66077"
}
```



---

archive/issue_comments_371373.json:
```json
{
    "body": "With #27008, #27009 and #27010, all remaining doctest errors in `graph.py` concern the order in which solutions are displayed (not always the same in py2 and py3).",
    "created_at": "2019-01-03T15:14:17Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371373",
    "user": "https://github.com/dcoudert"
}
```

With #27008, #27009 and #27010, all remaining doctest errors in `graph.py` concern the order in which solutions are displayed (not always the same in py2 and py3).



---

archive/issue_events_066078.json:
```json
{
    "actor": "https://github.com/embray",
    "created_at": "2019-01-15T18:15:21Z",
    "event": "demilestoned",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "milestone": "sage-8.6",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sagetest/issues/26403#event-66078"
}
```



---

archive/issue_events_066079.json:
```json
{
    "actor": "https://github.com/embray",
    "created_at": "2019-01-15T18:15:21Z",
    "event": "milestoned",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "milestone": "sage-8.7",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sagetest/issues/26403#event-66079"
}
```



---

archive/issue_comments_371374.json:
```json
{
    "body": "Retarging tickets optimistically to the next milestone.  If you are responsible for this ticket (either its reporter or owner) and don't believe you are likely to complete this ticket before the next release (8.7) please retarget this ticket's milestone to sage-pending or sage-wishlist.",
    "created_at": "2019-01-15T18:15:21Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371374",
    "user": "https://github.com/embray"
}
```

Retarging tickets optimistically to the next milestone.  If you are responsible for this ticket (either its reporter or owner) and don't believe you are likely to complete this ticket before the next release (8.7) please retarget this ticket's milestone to sage-pending or sage-wishlist.



---

archive/issue_comments_371375.json:
```json
{
    "body": "I would like to ask some questions about how to deal with simplicial complexes.\n\n- The `is_isomorphic` method for simplicial complexes translates each complex to a graph and then asks whether they are isomorphic, preserving edge labels. Some edges are labeled and some are not: the edge labels look like\n\n```\nsage: g.edge_labels()\n[None,\n None,\n 'special_edge',\n 'special_edge']\n```\n This fails in Python 3 because it tries to sort edges, and it can't sort `str` and `NoneType`. It is easy enough to provide labels in this case, but should the graph `is_isomorphic` method deal well with graphs where some edges have labels and some don't? (I don't have an opinion and I will fix the simplicial complex code separately.)\n\n- Again with the `is_isomorphic` method for simplicial complexes: it defines a graph with vertices labeled by the vertices of the complex and its facets, and it adds an edge between a vertex `v` and a facet `f` if `v` is in `f`. The graph code tries to sort the vertices, and since the vertices in a simplicial complex can be integers, strings, whatever, the sorting goes badly. Again, I think I can handle this for simplicial complexes by translating everything to something sortable, then sorting, and then translating back at the end, but should the graph `is_isomorphic` method handle these sorts of cases better?\n\n- Consider this:\n\n```\nsage: S1 = simplicial_complexes.Sphere(1)\nsage: g = S1.wedge(S1).graph()\nsage: g\nGraph on 5 vertices\nsage: g.vertices()\n[0, 'L1', 'L2', 'R1', 'R2']\n```\n In Python 3, the last line instead raises an error:\n\n```\n---------------------------------------------------------------------------\nTypeError                                 Traceback (most recent call last)\n<ipython-input-4-b2087dfc7b91> in <module>()\n----> 1 g.vertices()\n\n/Users/jpalmier/Desktop/Sage/sage_builds/PYTHON3/sage-8.6.rc1/local/lib/python3.6/site-packages/sage/graphs/generic_graph.py in vertices(self, sort, key)\n  10053             raise ValueError('sort keyword is False, yet a key function is given')\n  10054         if sort:\n> 10055             return sorted(list(self.vertex_iterator()), key=key)\n  10056         return list(self.vertex_iterator())\n  10057 \n\nTypeError: '<' not supported between instances of 'int' and 'str'\nsage: g.min_spanning_tree()\n---------------------------------------------------------------------------\nTypeError                                 Traceback (most recent call last)\n<ipython-input-5-8a025bbcbd58> in <module>()\n----> 1 g.min_spanning_tree()\n\n/Users/jpalmier/Desktop/Sage/sage_builds/PYTHON3/sage-8.6.rc1/local/lib/python3.6/site-packages/sage/graphs/generic_graph.py in min_spanning_tree(self, weight_function, algorithm, starting_vertex, check)\n   4231                 return min_spanning_tree(g,\n   4232                                          weight_function=wfunction_float,\n-> 4233                                          algorithm=algorithm.split(\"_\")[0])\n   4234 \n   4235         if algorithm == \"Prim_fringe\":\n\n/Users/jpalmier/Desktop/Sage/sage_builds/PYTHON3/sage-8.6.rc1/local/lib/python3.6/site-packages/sage/graphs/base/boost_graph.pyx in sage.graphs.base.boost_graph.min_spanning_tree (build/cythonized/sage/graphs/base/boost_graph.cpp:9345)()\n    604 \n    605 \n--> 606 cpdef min_spanning_tree(g,\n    607                         weight_function=None,\n    608                         algorithm='Kruskal'):\n\n/Users/jpalmier/Desktop/Sage/sage_builds/PYTHON3/sage-8.6.rc1/local/lib/python3.6/site-packages/sage/graphs/base/boost_graph.pyx in sage.graphs.base.boost_graph.min_spanning_tree (build/cythonized/sage/graphs/base/boost_graph.cpp:9100)()\n    719     else:\n    720         edges = [(int_to_vertex[<int> result[2*i]], int_to_vertex[<int> result[2*i+1]]) for i in range(n-1)]\n--> 721         return [(min(e[0],e[1]), max(e[0],e[1]), g.edge_label(e[0], e[1])) for e in edges]\n    722 \n    723 \n\nTypeError: '<' not supported between instances of 'int' and 'str'\n```\n In Python 2, `g.min_spanning_tree()` works, while in Python 3, it raises a similar `TypeError`:\n\n```\nsage: g.min_spanning_tree()\n---------------------------------------------------------------------------\nTypeError                                 Traceback (most recent call last)\n<ipython-input-5-8a025bbcbd58> in <module>()\n----> 1 g.min_spanning_tree()\n\n/Users/jpalmier/Desktop/Sage/sage_builds/PYTHON3/sage-8.6.rc1/local/lib/python3.6/site-packages/sage/graphs/generic_graph.py in min_spanning_tree(self, weight_function, algorithm, starting_vertex, check)\n   4231                 return min_spanning_tree(g,\n   4232                                          weight_function=wfunction_float,\n-> 4233                                          algorithm=algorithm.split(\"_\")[0])\n   4234 \n   4235         if algorithm == \"Prim_fringe\":\n\n/Users/jpalmier/Desktop/Sage/sage_builds/PYTHON3/sage-8.6.rc1/local/lib/python3.6/site-packages/sage/graphs/base/boost_graph.pyx in sage.graphs.base.boost_graph.min_spanning_tree (build/cythonized/sage/graphs/base/boost_graph.cpp:9345)()\n    604 \n    605 \n--> 606 cpdef min_spanning_tree(g,\n    607                         weight_function=None,\n    608                         algorithm='Kruskal'):\n\n/Users/jpalmier/Desktop/Sage/sage_builds/PYTHON3/sage-8.6.rc1/local/lib/python3.6/site-packages/sage/graphs/base/boost_graph.pyx in sage.graphs.base.boost_graph.min_spanning_tree (build/cythonized/sage/graphs/base/boost_graph.cpp:9100)()\n    719     else:\n    720         edges = [(int_to_vertex[<int> result[2*i]], int_to_vertex[<int> result[2*i+1]]) for i in range(n-1)]\n--> 721         return [(min(e[0],e[1]), max(e[0],e[1]), g.edge_label(e[0], e[1])) for e in edges]\n    722 \n    723 \n\nTypeError: '<' not supported between instances of 'int' and 'str'\n```\n Once again, I could probably get around this by translating all of the vertices to something sortable and then translating back, but I do think that the graph theory code should be able to handle graphs whose vertices are a combination of `int` and `str` and anything else. I would like to put in a request for fixing this particular problem.",
    "created_at": "2019-01-15T21:52:00Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371375",
    "user": "https://github.com/jhpalmieri"
}
```

I would like to ask some questions about how to deal with simplicial complexes.

- The `is_isomorphic` method for simplicial complexes translates each complex to a graph and then asks whether they are isomorphic, preserving edge labels. Some edges are labeled and some are not: the edge labels look like

```
sage: g.edge_labels()
[None,
 None,
 'special_edge',
 'special_edge']
```
 This fails in Python 3 because it tries to sort edges, and it can't sort `str` and `NoneType`. It is easy enough to provide labels in this case, but should the graph `is_isomorphic` method deal well with graphs where some edges have labels and some don't? (I don't have an opinion and I will fix the simplicial complex code separately.)

- Again with the `is_isomorphic` method for simplicial complexes: it defines a graph with vertices labeled by the vertices of the complex and its facets, and it adds an edge between a vertex `v` and a facet `f` if `v` is in `f`. The graph code tries to sort the vertices, and since the vertices in a simplicial complex can be integers, strings, whatever, the sorting goes badly. Again, I think I can handle this for simplicial complexes by translating everything to something sortable, then sorting, and then translating back at the end, but should the graph `is_isomorphic` method handle these sorts of cases better?

- Consider this:

```
sage: S1 = simplicial_complexes.Sphere(1)
sage: g = S1.wedge(S1).graph()
sage: g
Graph on 5 vertices
sage: g.vertices()
[0, 'L1', 'L2', 'R1', 'R2']
```
 In Python 3, the last line instead raises an error:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-b2087dfc7b91> in <module>()
----> 1 g.vertices()

/Users/jpalmier/Desktop/Sage/sage_builds/PYTHON3/sage-8.6.rc1/local/lib/python3.6/site-packages/sage/graphs/generic_graph.py in vertices(self, sort, key)
  10053             raise ValueError('sort keyword is False, yet a key function is given')
  10054         if sort:
> 10055             return sorted(list(self.vertex_iterator()), key=key)
  10056         return list(self.vertex_iterator())
  10057 

TypeError: '<' not supported between instances of 'int' and 'str'
sage: g.min_spanning_tree()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-8a025bbcbd58> in <module>()
----> 1 g.min_spanning_tree()

/Users/jpalmier/Desktop/Sage/sage_builds/PYTHON3/sage-8.6.rc1/local/lib/python3.6/site-packages/sage/graphs/generic_graph.py in min_spanning_tree(self, weight_function, algorithm, starting_vertex, check)
   4231                 return min_spanning_tree(g,
   4232                                          weight_function=wfunction_float,
-> 4233                                          algorithm=algorithm.split("_")[0])
   4234 
   4235         if algorithm == "Prim_fringe":

/Users/jpalmier/Desktop/Sage/sage_builds/PYTHON3/sage-8.6.rc1/local/lib/python3.6/site-packages/sage/graphs/base/boost_graph.pyx in sage.graphs.base.boost_graph.min_spanning_tree (build/cythonized/sage/graphs/base/boost_graph.cpp:9345)()
    604 
    605 
--> 606 cpdef min_spanning_tree(g,
    607                         weight_function=None,
    608                         algorithm='Kruskal'):

/Users/jpalmier/Desktop/Sage/sage_builds/PYTHON3/sage-8.6.rc1/local/lib/python3.6/site-packages/sage/graphs/base/boost_graph.pyx in sage.graphs.base.boost_graph.min_spanning_tree (build/cythonized/sage/graphs/base/boost_graph.cpp:9100)()
    719     else:
    720         edges = [(int_to_vertex[<int> result[2*i]], int_to_vertex[<int> result[2*i+1]]) for i in range(n-1)]
--> 721         return [(min(e[0],e[1]), max(e[0],e[1]), g.edge_label(e[0], e[1])) for e in edges]
    722 
    723 

TypeError: '<' not supported between instances of 'int' and 'str'
```
 In Python 2, `g.min_spanning_tree()` works, while in Python 3, it raises a similar `TypeError`:

```
sage: g.min_spanning_tree()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-8a025bbcbd58> in <module>()
----> 1 g.min_spanning_tree()

/Users/jpalmier/Desktop/Sage/sage_builds/PYTHON3/sage-8.6.rc1/local/lib/python3.6/site-packages/sage/graphs/generic_graph.py in min_spanning_tree(self, weight_function, algorithm, starting_vertex, check)
   4231                 return min_spanning_tree(g,
   4232                                          weight_function=wfunction_float,
-> 4233                                          algorithm=algorithm.split("_")[0])
   4234 
   4235         if algorithm == "Prim_fringe":

/Users/jpalmier/Desktop/Sage/sage_builds/PYTHON3/sage-8.6.rc1/local/lib/python3.6/site-packages/sage/graphs/base/boost_graph.pyx in sage.graphs.base.boost_graph.min_spanning_tree (build/cythonized/sage/graphs/base/boost_graph.cpp:9345)()
    604 
    605 
--> 606 cpdef min_spanning_tree(g,
    607                         weight_function=None,
    608                         algorithm='Kruskal'):

/Users/jpalmier/Desktop/Sage/sage_builds/PYTHON3/sage-8.6.rc1/local/lib/python3.6/site-packages/sage/graphs/base/boost_graph.pyx in sage.graphs.base.boost_graph.min_spanning_tree (build/cythonized/sage/graphs/base/boost_graph.cpp:9100)()
    719     else:
    720         edges = [(int_to_vertex[<int> result[2*i]], int_to_vertex[<int> result[2*i+1]]) for i in range(n-1)]
--> 721         return [(min(e[0],e[1]), max(e[0],e[1]), g.edge_label(e[0], e[1])) for e in edges]
    722 
    723 

TypeError: '<' not supported between instances of 'int' and 'str'
```
 Once again, I could probably get around this by translating all of the vertices to something sortable and then translating back, but I do think that the graph theory code should be able to handle graphs whose vertices are a combination of `int` and `str` and anything else. I would like to put in a request for fixing this particular problem.



---

archive/issue_comments_371376.json:
```json
{
    "body": "Maybe one of these (the second one, I think) was taken care of by #27027.",
    "created_at": "2019-01-15T22:15:05Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371376",
    "user": "https://github.com/jhpalmieri"
}
```

Maybe one of these (the second one, I think) was taken care of by #27027.



---

archive/issue_comments_371377.json:
```json
{
    "body": "See #27067 for simplicial complexes fixes for the first two items.",
    "created_at": "2019-01-15T23:01:14Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371377",
    "user": "https://github.com/jhpalmieri"
}
```

See #27067 for simplicial complexes fixes for the first two items.



---

archive/issue_comments_371378.json:
```json
{
    "body": "I fully agree that the graph theory code should be able to handle vertices of different types. We have done significant progresses recently, and many methods are no longer sorting vertices. But a lot remains to be done since a huge part of the code uses vertex ids instead of internal int labels. We have to find solutions for instance for listing edges as end points are currently sorted...\n\nNote that I'm not able to try your example with py3 as this `S1.wedge(S1).graph()` already raises an error.",
    "created_at": "2019-01-17T13:58:26Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371378",
    "user": "https://github.com/dcoudert"
}
```

I fully agree that the graph theory code should be able to handle vertices of different types. We have done significant progresses recently, and many methods are no longer sorting vertices. But a lot remains to be done since a huge part of the code uses vertex ids instead of internal int labels. We have to find solutions for instance for listing edges as end points are currently sorted...

Note that I'm not able to try your example with py3 as this `S1.wedge(S1).graph()` already raises an error.



---

archive/issue_events_066080.json:
```json
{
    "actor": "https://github.com/dcoudert",
    "created_at": "2019-01-17T13:58:26Z",
    "event": "demilestoned",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "milestone": "sage-8.7",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sagetest/issues/26403#event-66080"
}
```



---

archive/issue_events_066081.json:
```json
{
    "actor": "https://github.com/dcoudert",
    "created_at": "2019-01-17T13:58:26Z",
    "event": "milestoned",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "milestone": "sage-pending",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sagetest/issues/26403#event-66081"
}
```



---

archive/issue_comments_371379.json:
```json
{
    "body": "First, you have all done great work with graphs, and you've made a tremendous amount of progress. Partly my questions were because I don't know whether some of these things have been done in tickets which have not yet been merged, but I think that's not the case.\n\nSpeaking of things which have not yet been merged, you need #26966 to be able to do `S1.wedge(S1).graph()`. Sorry, I hadn't realized that when I posted my example.",
    "created_at": "2019-01-17T18:08:14Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371379",
    "user": "https://github.com/jhpalmieri"
}
```

First, you have all done great work with graphs, and you've made a tremendous amount of progress. Partly my questions were because I don't know whether some of these things have been done in tickets which have not yet been merged, but I think that's not the case.

Speaking of things which have not yet been merged, you need #26966 to be able to do `S1.wedge(S1).graph()`. Sorry, I hadn't realized that when I posted my example.



---

archive/issue_comments_371380.json:
```json
{
    "body": "Also, as I hope was clear in my questions, the first two are not critiques but honest questions: I don't know how Sage should handle those situations. The fact that things happen to work in Python 2 does not mean that they should automatically work in Python 3. For example, if you explicitly ask for isomorphisms preserving edge labels, an argument could be made that you should actually label all of the edges. The last item on my list was an explicit request, which it sounds like you agree with but also sounds like it take some work.",
    "created_at": "2019-01-17T18:17:57Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371380",
    "user": "https://github.com/jhpalmieri"
}
```

Also, as I hope was clear in my questions, the first two are not critiques but honest questions: I don't know how Sage should handle those situations. The fact that things happen to work in Python 2 does not mean that they should automatically work in Python 3. For example, if you explicitly ask for isomorphisms preserving edge labels, an argument could be made that you should actually label all of the edges. The last item on my list was an explicit request, which it sounds like you agree with but also sounds like it take some work.



---

archive/issue_comments_371381.json:
```json
{
    "body": "With #26819, #27176, #27179, #27180, #27181, #27183 py3 and #27184, we fix 42 doctests in `generic_graph.py` !!!\n\nThe remainging issues are in `relabel`, `is_isomorphic` and `graph_isom_equivalent_non_edge_labeled_graph`.",
    "created_at": "2019-01-31T14:09:44Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371381",
    "user": "https://github.com/dcoudert"
}
```

With #26819, #27176, #27179, #27180, #27181, #27183 py3 and #27184, we fix 42 doctests in `generic_graph.py` !!!

The remainging issues are in `relabel`, `is_isomorphic` and `graph_isom_equivalent_non_edge_labeled_graph`.



---

archive/issue_comments_371382.json:
```json
{
    "body": "We have now drastically reduced the dependency to `.vertices()` and `.edges()` in the graph module (except in doctests, but we can specify `sort=True/False`). \n\nWe must now identify the tasks to do and schedule the work.\n\nSome issues to take care of in the graph module:\n- #22349 Deprecate sorting of methods `.vertices()` and `.edges()`.\n  - methods using matrices: `adjacency_matrix`, `distance_matrix`, etc. For all these methods, we can now give as input an ordering of the vertices that will be used to order rows and columns. Currently, we use `.vertices()` by default. If we switch to `list(G)` by default, we have to check that it's not breaking algorithms using these matrices.\n  - methods using `.relabel()`. We must check that changing the default ordering of vertices/edges is safe. Should be ok, but we never know...\n  - methods for isomorphisms were we also have issues with labels (see e.g. #27232 although the case is certainly beyond what we should reasonably do).\n- Deprecate sorting of vertex labels in method `iterator_edges` of `base/sparse_graph.pyx` (so of the vertices of an edge). Deprecating this behavior is certainly a hard task. A first step could be to order the vertices according internal integer value, thus ensuring that `(u, v)` will always be ordered the same way.\n\nAnd of course, changes done here will certainly break many methods in other modules...",
    "created_at": "2019-02-23T10:33:11Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371382",
    "user": "https://github.com/dcoudert"
}
```

We have now drastically reduced the dependency to `.vertices()` and `.edges()` in the graph module (except in doctests, but we can specify `sort=True/False`). 

We must now identify the tasks to do and schedule the work.

Some issues to take care of in the graph module:
- #22349 Deprecate sorting of methods `.vertices()` and `.edges()`.
  - methods using matrices: `adjacency_matrix`, `distance_matrix`, etc. For all these methods, we can now give as input an ordering of the vertices that will be used to order rows and columns. Currently, we use `.vertices()` by default. If we switch to `list(G)` by default, we have to check that it's not breaking algorithms using these matrices.
  - methods using `.relabel()`. We must check that changing the default ordering of vertices/edges is safe. Should be ok, but we never know...
  - methods for isomorphisms were we also have issues with labels (see e.g. #27232 although the case is certainly beyond what we should reasonably do).
- Deprecate sorting of vertex labels in method `iterator_edges` of `base/sparse_graph.pyx` (so of the vertices of an edge). Deprecating this behavior is certainly a hard task. A first step could be to order the vertices according internal integer value, thus ensuring that `(u, v)` will always be ordered the same way.

And of course, changes done here will certainly break many methods in other modules...



---

archive/issue_comments_371383.json:
```json
{
    "body": "Proposal for an `EdgeView` for graphs in #27408.",
    "created_at": "2019-03-03T21:27:33Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371383",
    "user": "https://github.com/dcoudert"
}
```

Proposal for an `EdgeView` for graphs in #27408.



---

archive/issue_comments_371384.json:
```json
{
    "body": "I don't know if this is a good idea or not, but see #27452 for a branch which fixes the last doctest failures in src/homology. This bypasses an issue with graphs, as mentioned above:\n\n```\nsage: S1 = simplicial_complexes.Sphere(1)\nsage: g = S1.wedge(S1).graph()\nsage: g\nGraph on 5 vertices\nsage: g.vertices()\n...\nTypeError: '<' not supported between instances of 'int' and 'str'\n```\nThe branch at #27452 relabels the graph of a simplicial complex so that the vertices are sortable, before trying to compute its minimal spanning tree. Is this just sweeping things under the rug, or is it okay?",
    "created_at": "2019-03-10T00:20:10Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371384",
    "user": "https://github.com/jhpalmieri"
}
```

I don't know if this is a good idea or not, but see #27452 for a branch which fixes the last doctest failures in src/homology. This bypasses an issue with graphs, as mentioned above:

```
sage: S1 = simplicial_complexes.Sphere(1)
sage: g = S1.wedge(S1).graph()
sage: g
Graph on 5 vertices
sage: g.vertices()
...
TypeError: '<' not supported between instances of 'int' and 'str'
```
The branch at #27452 relabels the graph of a simplicial complex so that the vertices are sortable, before trying to compute its minimal spanning tree. Is this just sweeping things under the rug, or is it okay?



---

archive/issue_comments_371385.json:
```json
{
    "body": "I think it is one way forward. It does suffer in large cases because you have to make a full copy of the simplicial complex's graph, but I don't think it is simply sweeping it under the rug. Perhaps for that a user should know better than to have complicated labels for such tests?",
    "created_at": "2019-03-10T00:47:20Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371385",
    "user": "https://github.com/tscrim"
}
```

I think it is one way forward. It does suffer in large cases because you have to make a full copy of the simplicial complex's graph, but I don't think it is simply sweeping it under the rug. Perhaps for that a user should know better than to have complicated labels for such tests?



---

archive/issue_comments_371386.json:
```json
{
    "body": "An alternative is to (optionally) make spanning tree methods return a `Graph`, in which case there is no need to sort end vertices.",
    "created_at": "2019-03-10T10:54:31Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371386",
    "user": "https://github.com/dcoudert"
}
```

An alternative is to (optionally) make spanning tree methods return a `Graph`, in which case there is no need to sort end vertices.



---

archive/issue_comments_371387.json:
```json
{
    "body": "Replying to [comment:67 dcoudert]:\n> An alternative is to (optionally) make spanning tree methods return a `Graph`, in which case there is no need to sort end vertices.\n\n\nMy gut reaction was that seems heavy handed. It would make sense, but I suspect the reason we don't is because it would be too expensive (relatively) to return a `(Di)Graph`. I haven't looked into this (it is too late for me to do so tonight), but just a thought.",
    "created_at": "2019-03-10T12:31:18Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371387",
    "user": "https://github.com/tscrim"
}
```

Replying to [comment:67 dcoudert]:
> An alternative is to (optionally) make spanning tree methods return a `Graph`, in which case there is no need to sort end vertices.


My gut reaction was that seems heavy handed. It would make sense, but I suspect the reason we don't is because it would be too expensive (relatively) to return a `(Di)Graph`. I haven't looked into this (it is too late for me to do so tonight), but just a thought.



---

archive/issue_comments_371388.json:
```json
{
    "body": "Some doctests are failing with `bliss` (not related to #27571) \n\n```\nsage -t src/sage/graphs/generators/families.py\n**********************************************************************\nFile \"src/sage/graphs/generators/families.py\", line 3191, in sage.graphs.generators.families.MathonPseudocyclicStronglyRegularGraph\nFailed example:\n    G3x3.automorphism_group(algorithm=\"bliss\").order() # optional - bliss\nExpected:\n    27\nGot:\n    3\n**********************************************************************\nFile \"src/sage/graphs/generators/families.py\", line 3196, in sage.graphs.generators.families.MathonPseudocyclicStronglyRegularGraph\nFailed example:\n    G9.automorphism_group(algorithm=\"bliss\").order() # optional - bliss\nExpected:\n    9\nGot:\n    3\n```",
    "created_at": "2019-03-30T14:03:49Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371388",
    "user": "https://github.com/dcoudert"
}
```

Some doctests are failing with `bliss` (not related to #27571) 

```
sage -t src/sage/graphs/generators/families.py
**********************************************************************
File "src/sage/graphs/generators/families.py", line 3191, in sage.graphs.generators.families.MathonPseudocyclicStronglyRegularGraph
Failed example:
    G3x3.automorphism_group(algorithm="bliss").order() # optional - bliss
Expected:
    27
Got:
    3
**********************************************************************
File "src/sage/graphs/generators/families.py", line 3196, in sage.graphs.generators.families.MathonPseudocyclicStronglyRegularGraph
Failed example:
    G9.automorphism_group(algorithm="bliss").order() # optional - bliss
Expected:
    9
Got:
    3
```



---

archive/issue_comments_371389.json:
```json
{
    "body": "We are almost done. We have currently 3 open tickets for the remaining failing doctests.\n\nNeeds work (contributors are more than welcome):\n- #27435: failing doctest in `graph_database.py` with `interactive_query`.\n- #28108: 2 `ValueError?` in `graph_generators.py` doctests with `plantri` optional package (follow up of #27948) \n\nNeeds review / help / comment:\n- #27232: fix failing doctests in `is_isomorphic`, which are the last 8 remaining failing doctests in `generic_graph.py`",
    "created_at": "2019-07-10T11:10:35Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371389",
    "user": "https://github.com/dcoudert"
}
```

We are almost done. We have currently 3 open tickets for the remaining failing doctests.

Needs work (contributors are more than welcome):
- #27435: failing doctest in `graph_database.py` with `interactive_query`.
- #28108: 2 `ValueError?` in `graph_generators.py` doctests with `plantri` optional package (follow up of #27948) 

Needs review / help / comment:
- #27232: fix failing doctests in `is_isomorphic`, which are the last 8 remaining failing doctests in `generic_graph.py`



---

archive/issue_comments_371390.json:
```json
{
    "body": "Replying to [comment:75 dcoudert]:\n> We are almost done. We have currently 3 open tickets for the remaining failing doctests.\n> \n> Needs work (contributors are more than welcome):\n> - #27435: failing doctest in `graph_database.py` with `interactive_query`.\n\n\nFor this one, I would be tempted to label it `# py2` since it seems to rely on the legacy Sage notebook. I can open a ticket if that sounds like a valid approach.",
    "created_at": "2019-07-10T17:20:28Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371390",
    "user": "https://github.com/jhpalmieri"
}
```

Replying to [comment:75 dcoudert]:
> We are almost done. We have currently 3 open tickets for the remaining failing doctests.
> 
> Needs work (contributors are more than welcome):
> - #27435: failing doctest in `graph_database.py` with `interactive_query`.


For this one, I would be tempted to label it `# py2` since it seems to rely on the legacy Sage notebook. I can open a ticket if that sounds like a valid approach.



---

archive/issue_comments_371391.json:
```json
{
    "body": "> > - #27435: failing doctest in `graph_database.py` with `interactive_query`.\n \n> \n> For this one, I would be tempted to label it `# py2` since it seems to rely on the legacy Sage notebook. I can open a ticket if that sounds like a valid approach.\n\n\nIt is related to sage notebook, but doctests are failing only with py3... So I don't know which is the best approach. Should we try to make a little change to sagenb to make #27435 fixing the issue, although sagenb will be removed in the future (but when?) ?",
    "created_at": "2019-07-10T18:35:55Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371391",
    "user": "https://github.com/dcoudert"
}
```

> > - #27435: failing doctest in `graph_database.py` with `interactive_query`.
 
> 
> For this one, I would be tempted to label it `# py2` since it seems to rely on the legacy Sage notebook. I can open a ticket if that sounds like a valid approach.


It is related to sage notebook, but doctests are failing only with py3... So I don't know which is the best approach. Should we try to make a little change to sagenb to make #27435 fixing the issue, although sagenb will be removed in the future (but when?) ?



---

archive/issue_comments_371392.json:
```json
{
    "body": "I apologize, my comments really belong on #27435. I'll continue the discussion there.",
    "created_at": "2019-07-10T18:50:15Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371392",
    "user": "https://github.com/jhpalmieri"
}
```

I apologize, my comments really belong on #27435. I'll continue the discussion there.



---

archive/issue_comments_371393.json:
```json
{
    "body": "All doctests in the standard graphs module are now fixed ! Thanks to all of you for your great help.\n\nWe can now check all optional packages. For instance, `plantri` has 2 failing doctests (#28108).",
    "created_at": "2019-07-29T11:06:24Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371393",
    "user": "https://github.com/dcoudert"
}
```

All doctests in the standard graphs module are now fixed ! Thanks to all of you for your great help.

We can now check all optional packages. For instance, `plantri` has 2 failing doctests (#28108).



---

archive/issue_comments_371394.json:
```json
{
    "body": "The last failing doctests of the optional packages (benzene, bliss, buckygen, csdp, dot2tex and graphviz, gap_packages, igraph and python_igraph, mcqd, plantri, tdlib) are fixed with #27948, #28108  and #28371.",
    "created_at": "2019-08-20T14:11:58Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371394",
    "user": "https://github.com/dcoudert"
}
```

The last failing doctests of the optional packages (benzene, bliss, buckygen, csdp, dot2tex and graphviz, gap_packages, igraph and python_igraph, mcqd, plantri, tdlib) are fixed with #27948, #28108  and #28371.



---

archive/issue_comments_371395.json:
```json
{
    "body": "it's time to close this ticket.",
    "created_at": "2020-04-18T10:18:20Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371395",
    "user": "https://github.com/dcoudert"
}
```

it's time to close this ticket.



---

archive/issue_comments_371396.json:
```json
{
    "body": "Changing status from new to needs_review.",
    "created_at": "2020-04-18T10:18:20Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371396",
    "user": "https://github.com/dcoudert"
}
```

Changing status from new to needs_review.



---

archive/issue_comments_371397.json:
```json
{
    "body": "Changing status from needs_review to positive_review.",
    "created_at": "2020-04-18T17:43:35Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371397",
    "user": "https://github.com/jhpalmieri"
}
```

Changing status from needs_review to positive_review.



---

archive/issue_comments_371398.json:
```json
{
    "body": "Yes, I agree.",
    "created_at": "2020-04-18T17:43:35Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371398",
    "user": "https://github.com/jhpalmieri"
}
```

Yes, I agree.



---

archive/issue_events_066082.json:
```json
{
    "actor": "https://github.com/fchapoton",
    "created_at": "2020-04-20T16:41:53Z",
    "event": "demilestoned",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "milestone": "sage-pending",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sagetest/issues/26403#event-66082"
}
```



---

archive/issue_events_066083.json:
```json
{
    "actor": "https://github.com/fchapoton",
    "created_at": "2020-04-20T16:41:53Z",
    "event": "milestoned",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "milestone": "sage-duplicate/invalid/wontfix",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sagetest/issues/26403#event-66083"
}
```



---

archive/issue_events_066084.json:
```json
{
    "actor": "https://github.com/fchapoton",
    "created_at": "2020-06-20T06:35:55Z",
    "event": "closed",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sagetest/issues/26403#event-66084"
}
```



---

archive/issue_comments_371399.json:
```json
{
    "body": "Resolution: worksforme",
    "created_at": "2020-06-20T06:35:55Z",
    "issue": "https://github.com/sagemath/sagetest/issues/26403",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sagetest/issues/26403#issuecomment-371399",
    "user": "https://github.com/fchapoton"
}
```

Resolution: worksforme
